File: Leetcode/postorder_traversal.py
# https://leetcode.com/problems/binary-tree-postorder-traversal/
'''
    Author - Subhajit Das
    University of Engineering and Management, Kolkata
    13/10/2019
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        result = list()
        self.postorderTraversalRecursive(root, result)
        return result

    def postorderTraversalRecursive(self, root: TreeNode, result: list) -> None:
        if(root == None):
            return

        if(root.left != None):
            self.postorderTraversalRecursive(root.left, result)

        if(root.right != None):
            self.postorderTraversalRecursive(root.right, result)

        result.append(root.val)

File: InterviewBit/PostorderTraversal.java
// https://www.interviewbit.com/problems/postorder-traversal/
/*
    Author @Subhajit Das (sdsubhajitdas.github.io)
    SWE @Turbot HQ India PVT Ltd.
    14/04/2021
*/

import java.util.*;

/**
 * Definition for binary tree
 * class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) {
 *      val = x;
 *      left=null;
 *      right=null;
 *     }
 * }
 */
public class Solution {
    public ArrayList<Integer> postorderTraversal(TreeNode root) {
        ArrayList<Integer> resultPostorder = new ArrayList<>();
        this.postorderTraversalHelper(root, resultPostorder);
    }

    private void postorderTraversalHelper(TreeNode node, ArrayList<Integer> order) { 
        if(node == null) return;
        if(node.left == null && node.right == null) {
            order.add(node.val);
            return;
        }

        this.postorderTraversalHelper(node.left, order);
        this.postorderTraversalHelper(node.right, order);
        order.add(node.val);
    }
}
